### PR TITLE
Adds support for Tiered Caching

### LINKED ISSUE #2100: Add Support for Tiered Caching
### Current Terraform and Cloudflare provider version

Terraform v1.2.7
Provider v3.3.0

### Description

Adds support for tiered caching using generic and smart topologies.

### Use cases

Tiered Cache is a commonly supported use case within the Dashboard. This adds support for existing mechanisms through Terraform.

### Potential Terraform configuration

```hcl
resource "cloudflare_tiered_cache" "example" {
  zone_id = "abcd..."
  cache_type = "smart"
}
```

### References

_No response_

### PR DESCRIPTION
**Pull request by iveelsm**
This adds support for manipulation, creating and deletion of tiered cache settings through the existing mechanisms in the Cloudflare API. This adds support for settings generic and smart tiered cache topologies through the terraform provider.

Implementation:

* Create/Update, Delete and Get methods for Tiered Cache settings

Dependent on the merging of the cloudflare-go PR: https://github.com/cloudflare/cloudflare-go/pull/1149

Once that PR is merged, this require additional updates to pull in the latest published version.

Closes #2100

### CODE REVIEW

File: internal/provider/resource_cloudflare_tiered_cache.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	"context"
+	"fmt"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceCloudflareTieredCache() *schema.Resource {
+	return &schema.Resource{
```
Comment (jacobbednarz) [inline internal/provider/resource_cloudflare_tiered_cache.go:13]: we'll need to add a `Description` to this for documentation output.

**Comment by jacobbednarz** [comment]
couple of things:

- you'll need to also add this to the `provider.go` (see #1979 for an example) for this to actually work. the acceptance tests will fail as is at the moment even if you shim in cloudflare-go as the resource isn't defined.
- you can run `make docs` to generate the automatic documentation. this is also run on `master` merges but doing it now gives you a chance to check it all out before then.

other than that, looking good! i'll run the acceptance tests once the couple of things are addressed.

**Comment by iveelsm** [comment]
Word! Can do. I will probably get it fully ready once we merged the cloudflare-go PR.

**Comment by jacobbednarz** [comment]
if you'd like to work on them together, there is the option of doing https://github.com/cloudflare/terraform-provider-cloudflare/blob/master/contributing/development.md#pulling-in-unreleased-changes-from-go-library-cloudflare-go locally that allows you to get this 95% of the way there and just wait for the version bump to arrive in the mainline.

File: internal/provider/resource_cloudflare_tiered_cache.go (hunk 1 above)
Comment (iveelsm) [reply internal/provider/resource_cloudflare_tiered_cache.go:14]: Done!

**Comment by jacobbednarz** [comment]
acceptance tests all 🍏 

```
TF_ACC=1 go test $(go list ./...) -v -run "^TestCloudflareTieredCache_" -count 1 -parallel 1 -timeout 120m -parallel 1
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
=== RUN   TestCloudflareTieredCache_Smart
--- PASS: TestCloudflareTieredCache_Smart (7.22s)
=== RUN   TestCloudflareTieredCache_Generic
--- PASS: TestCloudflareTieredCache_Generic (6.30s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/provider	13.956s
```

**Pull request merged**